lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1534,10 +1555,11 @@ bool TypeChecker::coerceParameterListToType(ParameterList *P, ClosureExpr *CE,
                  diag::param_type_non_materializable_tuple, ty);
       }
     }
-    
-    if (param->isInvalid())
-      param->overwriteType(ErrorType::get(Context));
-    else
+
+    // If contextual type is invalid and we have a valid argument type
+    // trying to coerce argument to contextual type would mean erasing
+    // valuable diagnostic information.
+    if (isValidType(ty) || shouldOverwriteParam(param))
       param->overwriteType(ty);
```

Comment:
Previously we were always overwriting the type, and now we are doing so only sometimes.

Should this be reflected in some way in the result that we return?

We end up through this code from both CSApply and CSDiag and at a quick glance it's not clear what the implications are of not always overwriting the type here.